Improve libfuzzer integration

kyakdan opened the pull request.

kyakdan committed:
Make native libfuzzer hooks available in go-fuzz-dep

kyakdan committed:
Copy .h and .cpp files when copying go-fuzz-dep to instrument code

kyakdan committed:
Intercept string comparisons, integer comparisons, integer division, and indexing expressions

kyakdan committed:
Remove tracing of index expressions

kyakdan committed:
Fix creating saving defines for call expressions

kyakdan committed:
Fix signedness for integer hooks

kyakdan committed:
Adjust original expression after adding defines to handle expressions with potential side effects

kyakdan committed:
Do the conversions from integer hook IDs to pointers representing fake PCs in C.

Doing this in Go causes multiple cgo issues

kyakdan committed:
Fix default naming of the build output to always contain the name of the package containing the fuzz target

fmeum commented on go-fuzz-dep/sanitizer_hooks_with_pc.h:
```diff
@@ -0,0 +1,59 @@
+// gep hooks that accept an additional caller_pc argument that can be used to
+// pass a custom value that is recorded as the caller's instruction pointer
+// ("program counter"). This allows synthetic program counters obtained from
+// Java coverage information to be used with libFuzzer's value profile, with
```
This mentions Java.

kyakdan replied:
Changed.

fmeum commented on go-fuzz-dep/sanitizer_hooks_with_pc.cpp:
```diff
@@ -0,0 +1,215 @@
+
+// The first four registers to pass arguments in according to the
+// platform-specific x64 calling convention.
+#ifdef _WIN64
```
Before we submit this, we might want to port the inline asm to the one remaining relevant architecture, aarch64.

fmeum commented on go-fuzz-build/main.go:
```diff
@@ -605,7 +605,9 @@ func (c *Context) copyFuzzDep() {
 	c.mkdirAll(newDir)
 	dep := c.packageNamed("github.com/dvyukov/go-fuzz/go-fuzz-dep")
-	for _, f := range dep.GoFiles {
+	pkgFiles := dep.GoFiles
```
Why is this needed?

kyakdan replied:
I'm not sure I get the question. `go-fuzz-dep` package is used by the instrumented code and needs to be compiled. So, here we copy both the Go and C/C++ files into the work directory where the `GOPATH` environment is prepared.

fmeum replied:
I see, I missed that the trampoline was added to `go-fuzz-dep` and thus needs to be picked up here.

kyakdan committed:
Adjust comments in sanitizer_hooks_with_pc.h to mention Go

fmeum commented on go-fuzz-build/hooks.go:
```diff
@@ -0,0 +1,367 @@
+func hookId() *ast.BasicLit {
+	return &ast.BasicLit{
+		Kind:  token.INT,
+		Value: strconv.Itoa(int(rand.Uint32())),
```
It would be great if we could make the hook IDs reproducible. How about using an RNG seeded with a hash of the filename or contents?

kyakdan replied:
Good point. Done!

fmeum commented on go-fuzz-build/hooks.go:
```diff
@@ -0,0 +1,367 @@
+func isFunctionFromStringsPackage(n ast.Node) bool {
+	if call, ok := n.(*ast.CallExpr); ok {
+		if fnSel, ok2 := call.Fun.(*ast.SelectorExpr); ok2 {
+			if pkg, ok3 := fnSel.X.(*ast.Ident); ok3 && pkg.Name == "strings" {
```
Could this break with renamed imports or is `pkg` the canonical package name here? It should be rare to rename a package so that it clashes with a common stdlib package though, so may be safe to ignore.

kyakdan replied:
agreed!

fmeum commented on go-fuzz-build/hooks.go:
```diff
@@ -0,0 +1,367 @@
+
+var stringsHooks = map[string]string{
+	"Compare":   "HookStrcmp",
+	"EqualFold": "HookStrcmpBoolResult",
```
`__sanitizer_weak_hook_strcasecmp` exists, but just forwards to the regular `strcmp` hook, so it's probably not worth the effort to disambiguate them here.

kyakdan replied:
That is why I've only implemented the `__sanitizer_weak_hook_strcmp` hook

kyakdan committed:
Limit the scope of applying hooks to functions.

This caused problems with global const expressions

kyakdan committed:
Minor refactoring

kyakdan committed:
Make hook Ids reproducible.

Use the hash of the source file path as seed for generating hook Ids

fmeum commented on go-fuzz-build/hooks.go:
```diff
@@ -0,0 +1,387 @@
+// applyHooks applies the registered transformers on each visited AST node and replace that node
+// by the transformed one in case the returned node is not nil
+func (f *File) applyHooks() {
+	// seed PRNG based the hash of the filepath in order to get reproducible hook Ids
```
```suggestion
	// seed PRNG with the hash of the filepath in order to get reproducible hook Ids
```

kyakdan replied:
fixed as suggested

kyakdan committed:
Add support for libfuzzer's 8bit counters to avoid collisions caused by the fixed size of coverage map

kyakdan committed:
Remove the need to create a section for coverage counters

since we call __sanitizer_cov_8bit_counters_init, we don't need sections anymore. This gives us support for platforms other than Linux for free.

kyakdan committed:
fix comment

kyakdan committed:
Use LLVMFuzzerInitialize to initialize the 8bit counters

kyakdan committed:
Keep original constant coverage map for the sonar mode

kyakdan committed:
Consider a statement to potentially have a side effects if it contains a call expression or an expression statement

kyakdan committed:
Implement the NeverZero optimization proposed in AFL++

fmeum commented:
The new commits look good to me.

kyakdan committed:
Add include flag to go-fuzz-build

kyakdan committed:
Explicit error when failing to resolve a package
